File: go-webapp-arduino/api/factory/lmp-device-register.go
package factory

import (
	"bytes"
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/x509"
	"crypto/x509/pkix"
	"encoding/asn1"
	"encoding/json"
	"encoding/pem"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"path/filepath"
	"time"

	"github.com/google/uuid"
	log "github.com/inconshreveable/log15"
)

type AuthPrompt func(verificationUri, userCode string)

type DeviceCreateOpts struct {
	Factory       string
	OtaTag        string
	IsProd        bool
	HardwareId    string
	SotaConfigDir string
}

func NewFioDevice(opts DeviceCreateOpts, prompt AuthPrompt) error {
	deviceUuid := uuid.NewString()
	pkey, csr, err := genKey(opts.Factory, deviceUuid, opts.IsProd)
	if err != nil {
		return err
	}

	path := filepath.Join(opts.SotaConfigDir, "pkey.pem")
	if err = ioutil.WriteFile(path, pkey, 0o644); err != nil {
		return err
	}

	token, err := getAccessToken(opts.Factory, deviceUuid, prompt)
	if err != nil {
		return err
	}
	return createDevice(opts, deviceUuid, token, csr)
}

func genKey(factory, uuid string, production bool) ([]byte, []byte, error) {
	key, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	if err != nil {
		return nil, nil, err
	}

	subject := pkix.Name{
		CommonName:         uuid,
		OrganizationalUnit: []string{factory},
	}
	rawSubj := subject.ToRDNSequence()
	businessCategoryOid := asn1.ObjectIdentifier{2, 5, 4, 15}
	if production {
		rawSubj = append(rawSubj, []pkix.AttributeTypeAndValue{
			{Type: businessCategoryOid, Value: "production"},
		})
	}
	asn0Subj, _ := asn1.Marshal(rawSubj)

	digitalSignature, _ := asn1.Marshal(asn1.BitString{
		Bytes: []byte{128},
	})
	tlsWebClientAuth, _ := asn1.Marshal(
		[]asn1.ObjectIdentifier{asn1.ObjectIdentifier([]int{1, 3, 6, 1, 5, 5, 7, 3, 2})},
	)

	template := x509.CertificateRequest{
		RawSubject: asn0Subj,
		ExtraExtensions: []pkix.Extension{
			{
				Id:       asn1.ObjectIdentifier([]int{2, 5, 29, 15}),
				Critical: true,
				Value:    digitalSignature,
			},
			{
				Id:       asn1.ObjectIdentifier([]int{2, 5, 29, 37}),
				Critical: true,
				Value:    tlsWebClientAuth,
			},
		},
	}
	csr, err := x509.CreateCertificateRequest(rand.Reader, &template, key)
	if err != nil {
		return nil, nil, err
	}
	block := &pem.Block{
		Type:  "CERTIFICATE REQUEST",
		Bytes: csr,
	}
	csrPem := pem.EncodeToMemory(block)

	privBytes, err := x509.MarshalECPrivateKey(key)
	if err != nil {
		return nil, nil, err
	}
	block = &pem.Block{
		Type:  "EC PRIVATE KEY",
		Bytes: privBytes,
	}
	keyPem := pem.EncodeToMemory(block)

	return keyPem, csrPem, nil
}

func readResponse(res *http.Response) ([]byte, error) {
	defer res.Body.Close()
	return ioutil.ReadAll(res.Body)
}

func getAccessToken(factory, deviceUuid string, prompt AuthPrompt) (string, error) {
	form := url.Values{"client_id": {deviceUuid}}
	res, err := http.PostForm("https://app.foundries.io/oauth/authorization/device/", form)
	if err != nil {
		return "", err
	}
	body, err := readResponse(res)
	if err != nil {
		return "", err
	} else if res.StatusCode != 200 {
		return "", fmt.Errorf("HTTP_%d: %s", res.StatusCode, string(body))
	}

	type oauthClaim struct {
		DeviceCode      string `json:"device_code"`
		UserCode        string `json:"user_code"`
		VerificationUri string `json:"verification_uri"`
		Interval        int    `json:"interval"`
		ExpiresIn       int    `json:"expires_in"`
	}
	var claim oauthClaim
	if err = json.Unmarshal(body, &claim); err != nil {
		return "", err
	}

	prompt(claim.VerificationUri, claim.UserCode)

	form = url.Values{
		"grant_type":  {"urn:ietf:params:oauth:grant-type:device_code"},
		"device_code": {claim.DeviceCode},
		"client_id":   {deviceUuid},
		"scope":       {factory + ":devices:create"},
	}

	i := 0
	// WHEELS := []byte{'|', '/', '-', '\\'}
	for {
		res, err := http.PostForm("https://app.foundries.io/oauth/token/", form)
		if err != nil {
			log.Error("lmp-device-register getAccessToken", "Unable to check for token:", err)
		} else {
			body, err := readResponse(res)
			if err != nil {
				fmt.Println("Unable to read HTTP response:", err)
			} else if res.StatusCode == 200 {
				type Token struct {
					AccessToken string `json:"access_token"`
				}
				var token Token
				if err = json.Unmarshal(body, &token); err != nil {
					return "", err
				}
				// fmt.Println()
				return token.AccessToken, nil
			} else if res.StatusCode == 400 {
				// fmt.Printf("Waiting for authorization %c\r", WHEELS[i%len(WHEELS)])
				i += 1
			} else {
				return "", fmt.Errorf("HTTP_%d: %s", res.StatusCode, string(body))
			}
		}
		time.Sleep(time.Duration(claim.Interval) * time.Second)
	}
}

type PacmanOverrides struct {
	Type            string `json:"type"`
	Tag             string `json:"tags"`
	ComposeAppsRoot string `json:"compose_apps_root"`
	ComposeApps     string `json:"compose_apps,omitempty"`
}
type Overrides struct {
	Pacman PacmanOverrides `json:"pacman"`
}
type deviceCreateRequest struct {
	Name        string    `json:"name"`
	Uuid        string    `json:"uuid"`
	Csr         string    `json:"csr"`
	HardwareId  string    `json:"hardware-id"`
	SotaConfDir string    `json:"sota-config-dir"`
	Overrides   Overrides `json:"overrides"`
}

func createDevice(opts DeviceCreateOpts, uuid, token string, csr []byte) error {
	dc := deviceCreateRequest{
		Name:        uuid,
		Uuid:        uuid,
		Csr:         string(csr),
		HardwareId:  opts.HardwareId,
		SotaConfDir: opts.SotaConfigDir,
		Overrides: Overrides{
			Pacman: PacmanOverrides{
				Type:            "\"ostree+compose_apps\"",
				Tag:             "\"" + opts.OtaTag + "\"",
				ComposeAppsRoot: "\"" + filepath.Join(opts.SotaConfigDir, "compose-apps") + "\"",
			},
		},
	}

	data, err := json.Marshal(dc)
	if err != nil {
		return err
	}

	req, err := http.NewRequest("POST", "https://api.foundries.io/ota/devices/", bytes.NewBuffer(data))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json; charset=UTF-8")
	req.Header.Set("Authorization", "Bearer "+token)
	res, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}
	body, err := readResponse(res)
	if err != nil {
		return err
	} else if res.StatusCode != 201 {
		return fmt.Errorf("HTTP_%d: %s", res.StatusCode, string(body))
	}
	var files map[string]string
	if err = json.Unmarshal(body, &files); err != nil {
		return err
	}
	for name, content := range files {
		path := filepath.Join(opts.SotaConfigDir, name)
		if err = ioutil.WriteFile(path, []byte(content), 0o644); err != nil {
			return err
		}
	}
	return nil
}

// func main() {
// 	opts := DeviceCreateOpts{
// 		Factory:       "andy-corp",
// 		OtaTag:        "master",
// 		IsProd:        true,
// 		HardwareId:    "intel",
// 		SotaConfigDir: "/var/sota",
// 	}

// 	prompt := func(verificationUri, userCode string) {
// 		fmt.Printf("Please vist %s and enter \"%s\" as the user code\n", verificationUri, userCode)
// 	}
// 	err := NewFioDevice(opts, prompt)
// 	if err != nil {
// 		panic(err)
// 	}
// }

File: arduino-ootb-webapp/lmp_device_register.py
#!/usr/bin/python3

from argparse import ArgumentParser
import os
import subprocess
import sys
from tempfile import NamedTemporaryFile
from time import sleep
from typing import NamedTuple, Tuple
from uuid import uuid4

import requests


class Options(NamedTuple):
    factory: str
    sota_dir: str
    tag: str
    apps: str
    hwid: str
    uuid: str
    name: str
    production: bool


def create_key(uuid: str, factory: str, production: bool) -> Tuple[bytes, bytes]:
    key = subprocess.check_output(
        ["openssl", "ecparam", "-genkey", "-name", "prime256v1"]
    )

    with NamedTemporaryFile() as cnf:
        cnf.write(
            f"""[req]
prompt = no
distinguished_name = dn
req_extensions = ext

[dn]
CN={uuid}
OU={factory}
""".encode()
        )
        if production:
            cnf.write(b"businessCategory=production\n")
        cnf.write(
            b"""
[ext]
keyUsage=critical, digitalSignature
extendedKeyUsage=critical, clientAuth
"""
        )
        cnf.flush()

        r = subprocess.run(
            ["openssl", "req", "-new", "-config", cnf.name, "-key", "-"],
            input=key,
            stdout=subprocess.PIPE,
        )
        r.check_returncode()
    print("ANDY", r.stdout.decode())
    return key, r.stdout


def _get_oauth_token(factory: str, uuid: str):
    data = {"client_id": uuid}
    r = requests.post("https://app.foundries.io/oauth/authorization/device/", data=data)
    if r.status_code != 200:
        sys.exit(f"ERROR: HTTP_{r.status_code}: {r.text}")

    user_code = r.json()["user_code"]
    device_code = r.json()["device_code"]
    link = r.json()["verification_uri"]
    interval = r.json()["interval"]

    print(f"Active at {link} with code {user_code}")

    data = {
        "grant_type": "urn:ietf:params:oauth:grant-type:device_code",
        "device_code": r.json()["device_code"],
        "client_id": uuid,
        "scope": factory + ":devices:create",
    }

    i = 0
    WHEELS = ["|", "/", "-", "\\"]
    while True:
        r = requests.post("https://app.foundries.io/oauth/token/", data=data)
        if r.status_code == 200:
            return r.json()["access_token"]
        if r.status_code == 400:
            if r.json().get("error") == "authorization_pending":
                sys.stdout.write(
                    "Waiting for authorization " + WHEELS[i % len(WHEELS)] + "\r"
                )
                sys.stdout.flush()
                i += 1
            else:
                sys.exit("Error authorizing device: " + r.json()["error_description"])
        else:
            print(f"ERROR: HTTP_{r.status_code}: {r.text}")
        sleep(interval)


def main(opts: Options):
    token = _get_oauth_token(opts.factory, opts.uuid)
    pkey, csr = create_key(opts.uuid, opts.factory, opts.production)
    data = {
        "name": opts.name,
        "uuid": opts.uuid,
        "csr": csr.decode(),
        "hardware-id": opts.hwid,
        "sota-config-dir": opts.sota_dir,
        "overrides": {
            "pacman": {
                "type": '"ostree+compose_apps"',
                "tags": '"' + opts.tag + '"',
                "compose_apps_root": os.path.join(opts.sota_dir, "compose-apps"),
            },
        },
    }

    if opts.apps:
        data["overrides"]["pacman"]["compose_apps"] = '"' + opts.apps + '"'  # type: ignore

    r = requests.post(
        "https://api.foundries.io/ota/devices/",
        json=data,
        headers={"Authorization": f"Bearer {token}"},
    )
    if r.status_code != 201:
        sys.exit("ERROR: HTTP_%d: %s" % (r.status_code, r.text))

    with open(os.path.join(opts.sota_dir, "pkey.pem"), "wb") as f:
        f.write(pkey)

    for k, v in r.json().items():
        with open(os.path.join(opts.sota_dir, k), mode="w") as f:  # type: ignore
            f.write(v)

    with open(os.path.join(opts.sota_dir, "curl"), "wb") as f:
        os.fchmod(f.fileno(), 0o755)
        f.write(
            f"""#!/bin/sh -e
cd {opts.sota_dir}
curl --cert client.pem --key pkey.pem --cacert root.crt $*""".encode()
        )


def get_parser() -> ArgumentParser:
    p = ArgumentParser(description="Example client to ease local testing")
    p.add_argument(
        "--factory", "-f", required=True, help="Name of factory to register device in"
    )
    p.add_argument("--production", action="store_true", help="Make 'production' cert")
    p.add_argument("--sota-dir", "-d", default="/var/sota", help="default=%(default)s")
    p.add_argument("--tag", "-t", default="master", help="default=%(default)s")
    p.add_argument("--apps", "-a")
    p.add_argument(
        "--hwid", "-i", default="intel-corei7-64", help="default=%(default)s"
    )
    p.add_argument("--uuid", "-u", help="Default value is a random uuid")
    p.add_argument("--name", "-n", help="Default value is the uuid")
    return p


if __name__ == "__main__":
    args = get_parser().parse_args()
    if not args.uuid:
        args.uuid = str(uuid4())
    if not args.name:
        args.name = args.uuid
    args.sota_dir = os.path.abspath(args.sota_dir)

    options = Options(
        factory=args.factory,
        sota_dir=args.sota_dir,
        tag=args.tag,
        apps=args.apps,
        hwid=args.hwid,
        uuid=args.uuid,
        name=args.name,
        production=args.production,
    )
    main(options)
